SCA command control functions

jsturdy opened the pull request:
## Description

Initial placeholders for SCA interface functionality, as discussed in #93 

* Add local-only functions to control the use of the various interfaces:
  * Generic `sendSCACommand` and then interfaces to the different blocks that call this command
    * `sendSCACommandWithReply` for transactions that expect returned data
  * SCA `CTRL` interface
    * Generic control registers B, C, and D
    * Chip ID and SEU counters
  * SCA `ADC` interface placeholder
  * SCA `GPIO` interface placeholder
  * SCA `I2C` interface placeholder
* Add enums
  * Per interface
    * for the different channels (e.g., ADC channels)
    * for the different commands
  * Different SCA communication channels
  * Different bits in the SCA command error word

### Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Motivation and Context
Need to be able to reliably interface with the various interfaces on the SCA for GPIO/ADC/I2C

## How Has This Been Tested?
Has not been, for now just placeholders and framework for future development.

mexanick commented on include/amc/sca.h:
```diff
@@ -8,12 +8,102 @@
+
 /*** CTRL submodule ***/
 /*!
  * \brief Reset the SCA module
```
Please update doc with parameters

mexanick commented on include/amc/sca_enums.h:
```diff
@@ -0,0 +1,269 @@
+    enum ESCAResetType {
+      SCAReset  = 0x0, ///< Reset the SCA
+      HardReset = 0x1, ///< Use the SCA to send a TTC-style HardReset command
+    } SCAResetTupe;
```
`SCAResetTupe`-->`SCAResetType`

mexanick commented on include/amc/sca_enums.h:
```diff
@@ -0,0 +1,269 @@
+    } CTRLCommand;
+  };
+
+  struct I2CCommand {  //I2CCommand settings
```
Missed doc

mexanick commented on include/amc/sca_enums.h:
```diff
@@ -0,0 +1,269 @@
+    } GPIOCommand;
+  };
+
+  struct ADCCommand {  //ADCCommand settings
```
Missed doc

mexanick commented on include/amc/sca_enums.h:
```diff
@@ -0,0 +1,269 @@
+    } ADCCommand;
+  };  // struct ADCCommand
+
+      /**
```
Indentation should be fixed

jsturdy commented:
@mexanick, also rebased and I think all your comments addressed.
Would still like a cursory review from @evka85 to ensure that the framework makes sense from the FW perspective

evka85 commented:
This looks very nice! Although it would be nice to introduce a few most common high level functions like:
  * read all ADC channels
  * send a manual hard reset to a single chamber

jsturdy commented:
> This looks very nice! Although it would be nice to introduce a few most common high level functions like:

Absolutely, this was just to have the necessary components before handing the expansion task off.
Just wanted to have the core set up in the "right" way, i.e., if certain actions need to happen for **every** type of transaction, putting those in the wrappers (e.g., "disable monitoring")

evka85 commented:
Oh.. yeah disable monitoring I think should just always be done, so maybe even put it in sendScaCommand function. But if this way of ADC readout tests ok (which it does on GE2/1), then I would like to take the firmware based ADC monitoring out completely, because its potential to interfere with other functions as well as non-portability between GE1/1 and GE2/1..

Actually after taking a second look, I think you overlooked setting the GEM_AMC.SLOW_CONTROL.SCA.MANUAL_CONTROL.LINK_ENABLE_MASK, which defines to which OHs the command will go. It's a bitmask, I think you should just write linkMask to it on sendSCACommand()

jsturdy commented:
@evka85, OK, I added the monitoring on/off at the level of the wrappers (one layer abstracted from the `sendSCACommand` to minimize the number of times it is toggled)

I also added the per-link setting of the module reset, but in the python tool this is protected with a "check if the register exists", but perhaps this is only relevant for old firmware?
If it needs to be protected, that can be added by the next person.
